polkadot/runtime/common/src/paras_registrar/mod.rs

```diff
@@ -692,6 +906,138 @@ impl<T: Config> OnNewHead for Pallet<T> {
+		let new_deposit = T::ParaDeposit::get()
+			.saturating_add(per_byte_fee.saturating_mul((head.0.len() as u32).into()))
+			.saturating_add(per_byte_fee.saturating_mul((new_code.0.len() as u32).into()));
+
+		let current_deposit = info.deposit;
+
+		let free_upgrade = requirements == UpgradeRequirements::SkipRequirements ||
+			para.is_system() ||
+			lease_holding;
+
+		match (free_upgrade, info.billing_account.clone()) {
+			(false, Some(billing_account)) => {
+				if let Err(_) = <T as Config>::Currency::withdraw(
+					&billing_account,
+					T::UpgradeFee::get(),
+					WithdrawReasons::FEE,
+					ExistenceRequirement::KeepAlive,
+				) {
+					Self::deposit_event(Event::<T>::CodeUpgradeScheduleFailed(
+						CodeUpgradeScheduleError::FailedToPayUpgradeFee,
+					));
+					// This is an overestimate of the used weight, but it's better to be safe than
+					// sorry.
+					return Err(<T as Config>::WeightInfo::pre_code_upgrade())
+				}
+
+				let additional_deposit = new_deposit.saturating_sub(current_deposit);
+				if let Err(_) =
+					<T as Config>::Currency::reserve(&billing_account, additional_deposit)
+				{
+					Self::deposit_event(Event::<T>::CodeUpgradeScheduleFailed(
+						CodeUpgradeScheduleError::FailedToReserveDeposit,
+					));
+					// This is an overestimate of the used weight, but it's better to be safe than
+					// sorry.
+					return Err(<T as Config>::WeightInfo::pre_code_upgrade())
+				}
+
+				// Update the deposit to the new appropriate amount.
+				info.deposit = new_deposit;
+			},
+			_ => {
+				// No upgrade costs are required.
+			},
+		}
+
+		if current_deposit > new_deposit {
+			// The billing account should receive a refund if the current deposit exceeds the new
+			// required deposit, even if they did not initiate the upgrade.
+			//
+			// The excess deposit will be refunded to the caller upon the success of the code
+			// upgrade.
+			//
+			// The reason why the deposit is not instantly refunded is that scheduling a code
+			// upgrade doesn't guarantee the success of it.
+			//
+			// If we returned the deposit here, a possible attack scenario would be to register
+			// the validation code of a parachain and then schedule a code upgrade to set the
+			// code to an empty blob. In such a case, the pre-checking process would fail, so
+			// the old code would remain on-chain even though there is no deposit to cover it.
+
+			info.pending_deposit_refund = Some(current_deposit.saturating_sub(new_deposit));
+		}
+
+		Paras::<T>::insert(para, info);
+		Ok(<T as Config>::WeightInfo::pre_code_upgrade())
+	}
+}
+
+impl<T: Config> OnCodeUpgrade for Pallet<T> {
+	fn on_code_upgrade(id: ParaId) -> Weight {
+		let maybe_info = Paras::<T>::get(id);
+		if maybe_info.is_none() {
+			return T::DbWeight::get().reads(1)
+		}
+
+		let mut info = maybe_info
+			.expect("Ensured above that the deposit info is stored for the parachain; qed");
+
+		if let Some(rebate) = info.pending_deposit_refund {
```

Comment:
Seems like we are only refunding if the billing_account was set anyways, but then why bother setting `pending_deposit_refund` in that case?

Reply:
https://github.com/paritytech/polkadot-sdk/pull/2372#discussion_r1426235943